src/packages/cli/src/detach.ts

```diff
@@ -0,0 +1,259 @@
+import { fork } from "child_process";
+import createInstanceName from "./process-name";
+import envPaths from "env-paths";
+import psList from "@trufflesuite/ps-list";
+import {
+  existsSync,
+  mkdirSync,
+  rmSync,
+  writeFileSync,
+  readdirSync,
+  readFileSync
+} from "fs";
+import path from "path";
+import { StartArgs } from "./types";
+import { FlavorName } from "@ganache/flavors";
+
+export type DetachedInstance = {
+  instanceName: string;
+  pid: number;
+  startTime: number;
+  host: string;
+  port: number;
+  flavor: FlavorName;
+  cmd: string;
+  version: string;
+};
+
+const dataPath = envPaths(`Ganache/instances`).data;
+if (!existsSync(dataPath)) mkdirSync(dataPath, { recursive: true });
+
+const READY_MESSAGE = "ready";
+
+const START_ERROR = "An error ocurred spawning a detached instance of Ganache:";
+
+/**
+ * Notify that the detached instance has started and is ready to receive requests.
+ */
+export function notifyDetachedInstanceReady() {
+  // in "detach" mode, the parent will wait until the "ready" message is
+  // received before disconnecting from the child process.
+  process.send(READY_MESSAGE);
+}
+
+/**
+ * Attempt to find and remove the instance file for a detached instance.
+ * @param  {number} pid the pid of the detached instance
+ * @returns boolean indicating whether the instance file was cleaned up successfully
+ */
+export function removeDetachedInstanceFile(pid: number): boolean {
+  const instanceFilename = `${dataPath}/${pid}`;
+  if (existsSync(instanceFilename)) {
+    rmSync(instanceFilename);
```

Comment:
File system access is weird since time passes between the call to `existsSync` and `rmSync`, you'll still need to `try`/`catch` around `rmSync`.

Since you need to `try`/`catch` `rmSync`, there is no reason check `existsSync`.

In any case, if you are going to use `rmSync` you should make use of the useful options it gives you (like `force` and `retries`). And you should also probably use the async version, `rm`. If you don't want to use any of the `rm*` options, `unlink` is probably better, as it prevents the accidental deletion of directories -- which is always nice, even if it's not possible here (is it? :sweat_smile:).

Reply:
I've try/catch'd the rmSync call - I will look into opportunities to improve performance through async fs access and tie the conversation to your question "Why do you use synchronous methods instead of the async counterparts?".